src/Backups/getBackupDataFileNameGenerator.cpp

```diff
@@ -0,0 +1,109 @@
+#include <Backups/BackupFileInfo.h>
+#include <Backups/IBackupDataFileNameGenerator.h>
+#include <Backups/getBackupDataFileNameGenerator.h>
+
+#include <Common/ErrorCodes.h>
+#include <Common/Exception.h>
+
+#include <filesystem>
+#include <base/hex.h>
+
+#include <cstddef>
+
+namespace fs = std::filesystem;
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+
+extern const int ARGUMENT_OUT_OF_BOUND;
+extern const int BAD_ARGUMENTS;
+extern const int LOGICAL_ERROR;
+
+}
+
+}
+
+namespace
+{
+constexpr size_t DEFAULT_BACKUP_PREFIX_LENGTH = 0;
+
+class BackupDataFileNameGeneratorWithChecksumHexPrefix : public DB::IBackupDataFileNameGenerator
+{
+public:
+    explicit BackupDataFileNameGeneratorWithChecksumHexPrefix(size_t prefix_length_)
+        : prefix_length(prefix_length_)
+    {
+        if (prefix_length > 32)
+            throw DB::Exception(DB::ErrorCodes::ARGUMENT_OUT_OF_BOUND, "Prefix length {} out of bound", prefix_length);
+    }
+
+    std::string getName() const override { return std::string("checksum_hex_") + std::to_string(prefix_length); }
+
+    std::string generate(const DB::BackupFileInfo & file_info) override
+    {
+        if (file_info.file_name.empty())
+            throw DB::Exception(DB::ErrorCodes::LOGICAL_ERROR, "Backup file name should not be empty");
+
+        if (file_info.checksum == 0)
+            throw DB::Exception(DB::ErrorCodes::LOGICAL_ERROR, "Backup checksum should not be zero {}", file_info.file_name);
+
+        auto checksum_hex = getHexUIntLowercase(file_info.checksum);
+
+        return fs::path(checksum_hex.substr(0, prefix_length)) / checksum_hex.substr(prefix_length);
+    }
+
+private:
+    const size_t prefix_length;
+};
+
+class BackupDataFileNameGeneratorDefault : public DB::IBackupDataFileNameGenerator
```

Comment:
I suppose it's better to write something meaningful instead of `Default`. What do you think about names (for example) `BackupDataFileNameGeneratorFromChecksum` & `BackupDataFileNameGeneratorFromFirstFileName`?